src/components/join-flow/join-flow.jsx

```diff
@@ -61,59 +64,113 @@ class JoinFlow extends React.Component {
-                        // body can include zero or more error objects, each
-                        // with its own key and description. Here we assemble
-                        // all of them into a single string, errStr.
-                        const errorKeys = Object.keys(body[0].errors);
-                        errorKeys.forEach(key => {
-                            const val = body[0].errors[key];
-                            if (val && val[0]) {
-                                if (errStr.length) errStr += '; ';
-                                errStr += `${key}: ${val[0]}`;
-                            }
-                        });
-                    }
-                    if (!errStr.length && body[0].msg) errStr = body[0].msg;
+            const success = this.registrationIsSuccessful(err, body, res);
+            if (success) {
+                this.props.refreshSession();
+                this.setState({step: this.state.step + 1});
+                return;
+            }
+            // now we know something went wrong -- either an actual error (client-side
+            // or server-side), or just a problem with the registration content.
+
+            // if an actual error, prompt user to try again.
+            if (err || res.statusCode !== 200) {
+                this.setState({registrationError: {errorAllowsTryAgain: true}});
+                return;
+            }
+
+            // now we know there was a problem with the registration content.
+            // If the server provided us info on why registration failed,
+            // build a summary explanation string
+            let errorMsg = null;
+            const errorsFromResponse = this.getErrorsFromResponse(err, body, res);
+            // if there was exactly one error, check if we have a pre-written message
+            // about that precise error
+            if (errorsFromResponse.length === 1) {
+                const singleErrMsgId = validate.responseErrorMsg(
+                    errorsFromResponse[0].fieldName,
+                    errorsFromResponse[0].errorStr
+                );
+                if (singleErrMsgId) { // one error that we have a predefined explanation string for
+                    errorMsg = this.props.intl.formatMessage({id: singleErrMsgId});
                 }
             }
+            // if we have more than one error, build a custom message with all of the
+            // server-provided error messages
+            if (!errorMsg && errorsFromResponse.length > 0) {
+                errorMsg = this.getCustomErrMsg(errorsFromResponse);
+            }
             this.setState({
-                registrationError: errStr ||
-                    `${this.props.intl.formatMessage({
-                        id: 'registration.generalError'
-                    })} (${res.statusCode})`
+                registrationError: {
+                    errorAllowsTryAgain: false,
+                    errorMsg: errorMsg
```

Comment:
What happens when errorMsg is null? As far as I can tell this eventually gets sent to the RegistrationError component but it doesn't check whether the message exists before it tries to do something with it.  I may have just lost the trace of it though.  It might also be worth adding a test case for that if there isn't one already.

Reply:
registrationErrorStep's render() function has the line

```
                {this.props.errorMsg && (
```
...which tests for errorMsg's truthiness.

This approach is simpler than conditionally including or excluding the prop when we set the registrationError object.